tools/buildsys/src/args.rs

```diff
@@ -0,0 +1,214 @@
+These structs provide the CLI interface for buildsys which is called from Cargo.toml and accepts all
+of its input arguments from environment variables.
+
+!*/
+
+use buildsys::manifest::SupportedArch;
+use clap::{Parser, Subcommand};
+use std::path::PathBuf;
+use url::Url;
+
+/// A list of environment variables and the type of build that should be rerun if that environment
+/// variable changes. The build type is represented with bit flags so that we can easily list
+/// multiple build types for a single variable. See `[BuildType]` and `[rerun_for_envs]` below to
+/// see how this list is used.
+const REBUILD_VARS: [(&str, u8); 13] = [
+    ("BUILDSYS_ARCH", PACKAGE | VARIANT),
+    ("BUILDSYS_NAME", VARIANT),
+    ("BUILDSYS_OUTPUT_DIR", VARIANT),
+    ("BUILDSYS_PACKAGES_DIR", PACKAGE),
+    ("BUILDSYS_PRETTY_NAME", VARIANT),
+    ("BUILDSYS_ROOT_DIR", PACKAGE | VARIANT),
+    ("BUILDSYS_STATE_DIR", PACKAGE | VARIANT),
+    ("BUILDSYS_TIMESTAMP", VARIANT),
+    ("BUILDSYS_VARIANT", VARIANT),
+    ("BUILDSYS_VERSION_BUILD", VARIANT),
+    ("BUILDSYS_VERSION_IMAGE", VARIANT),
+    ("TLPRIVATE_SDK_IMAGE", PACKAGE | VARIANT),
+    ("TLPRIVATE_TOOLCHAIN", PACKAGE | VARIANT),
+];
+
+/// A tool for building Bottlerocket images and artifacts.
+#[derive(Debug, Parser)]
+pub(crate) struct Buildsys {
+    #[command(subcommand)]
+    pub(crate) command: Command,
+}
+
+#[derive(Subcommand, Debug)]
+pub(crate) enum Command {
+    BuildPackage(Box<BuildPackageArgs>),
+    BuildVariant(Box<BuildVariantArgs>),
+}
+
+impl Command {
+    pub(crate) fn build_type(&self) -> BuildType {
+        match self {
+            Command::BuildPackage(_) => BuildType::Package,
+            Command::BuildVariant(_) => BuildType::Variant,
+        }
+    }
+}
+
+/// Arguments common to all subcommands.
+#[derive(Debug, Parser)]
+pub(crate) struct Common {
+    #[arg(long, env = "BUILDSYS_ARCH")]
+    pub(crate) arch: SupportedArch,
+
+    #[arg(long, env = "BUILDSYS_OUTPUT_DIR")]
+    pub(crate) image_arch_variant_dir: PathBuf,
+
+    #[arg(long, env = "BUILDSYS_ROOT_DIR")]
+    pub(crate) root_dir: PathBuf,
+
+    #[arg(long, env = "BUILDSYS_STATE_DIR")]
+    pub(crate) state_dir: PathBuf,
+
+    #[arg(long, env = "BUILDSYS_TIMESTAMP")]
+    pub(crate) timestamp: String,
+
+    #[arg(long, env = "BUILDSYS_VERSION_FULL")]
+    pub(crate) version_full: String,
+
+    #[arg(long, env = "CARGO_MANIFEST_DIR")]
+    pub(crate) cargo_manifest_dir: PathBuf,
+
+    #[arg(long, env = "TLPRIVATE_SDK_IMAGE")]
+    pub(crate) sdk_image: String,
+
+    #[arg(long, env = "TLPRIVATE_TOOLCHAIN")]
+    pub(crate) toolchain: String,
+
+    #[arg(long, env = "TWOLITER_TOOLS_DIR")]
+    pub(crate) tools_dir: PathBuf,
+}
+
+/// Build RPMs from a spec file and sources.
+#[derive(Debug, Parser)]
+pub(crate) struct BuildPackageArgs {
+    #[arg(long, env = "BUILDSYS_PACKAGES_DIR")]
+    pub(crate) packages_dir: PathBuf,
+
+    #[arg(long, env = "BUILDSYS_VARIANT")]
+    pub(crate) variant: String,
```

Comment:
To check my understanding - we don't want this in `Common` because kits won't need this variable?

Reply:
> To check my understanding - we don't want this in `Common` because kits won't need this variable?

That's the idea, yeah. And I think our goal should be that package builds also don't need to know the variant name.